library/psa_crypto.c

```diff
@@ -981,18 +981,23 @@ psa_status_t psa_remove_key_data_from_memory(psa_key_slot_t *slot)
  * Persistent storage is not affected. */
 psa_status_t psa_wipe_key_slot(psa_key_slot_t *slot)
 {
+    if (slot->state != PSA_SLOT_PENDING_DELETION) {
```

Comment:
It can do, I think this change should be made. My original thinking was that by forcing the state from `PSA_SLOT_FULL` to `PSA_SLOT_PENDING_DELETION` we allow ourselves to memset the contents of the slot (except for the state variable) without holding the mutex. But after further thought that is probably (I am not very knowledgable about the efficiency of these operations) less efficient anyway, the current implementation would give a sequence of:

1. Lock mutex, set the state to pending deletion.
2. Unlock mutex, Call `psa_wipe_key_slot`.
3. Lock mutex, check state and reader count.
4. Unlock mutex, wipe slot contents.
5. Lock mutex, set the state to empty.
6. Unlock mutex, return.

Whereas if we allow `psa_wipe_key_slot` to accept full slots we need only lock the mutex, call psa_wipe_key_slot, then unlock the mutex. More operations are performed under the lock, but there is one lock call instead of three.

Reply:
My reasons for suggesting this are not so much for performance, but more like for maintainability (most of the deletion part of the state machine is at the same place in `psa_wipe_key_slot`) and simplicity (less code to add, less state transitions and `PSA_SLOT_PENDING_DELETION` only used where absolutely needed).

Regarding performance, I think we probably should only take the lock in the caller, do the checks, potential transitions, registering reads, calling `psa_wipe_key_slot` etc. Given the cost of taking a lock I think we can hold the lock across several operations or function calls. I think as long as we don't do crypto operations or I/O while holding the lock, we should be fine.